R/statistic-covariance.R

```diff
@@ -30,77 +8,56 @@ coefficientRelease <- function(formula, release, n) {
 #'    intercept should be added prior to evaluating the inner product x'x.
 #' @return The sensitivity of the data frame for which the covariance matrix
 #'   is being calculated.
+#'   
+#  A complete derivation of the sensitivity of the covariance may be found in 
+#' /extra_docs/sensitivities/covariance_sensitivity.pdf
 covarianceSensitivity <- function(n, rng, intercept) {
-  diffs <- apply(rng, 1, diff)
-  if (intercept) { diffs <- c(1, diffs) }
-  sensitivity <- c()
-  for (i in 1:length(diffs)) {
-    for (j in i:length(diffs)) {
-      s <- ((n - 1) / n) * diffs[i] * diffs[j]
-      sensitivity <- c(sensitivity, s)
+    diffs <- sapply(rng, diff)
+    if (intercept) { diffs <- c(1, diffs) }
+    sensitivity <- c()
+    c <- 2*(n-1)/n
```

Comment:
Overwriting `c` can lead to unexpected behavior.
Also, why the special-casing on the intercept? If a user augments a column of ones, then the bounds will be (1, 1), giving zero sensitivity as expected. This is also hardcoding the location of the intercept column.

Reply:
This is a feature that has existed since before I started working on the code; when the user initializes the statistic, they input a Boolean value for `intercept` and if it is `TRUE`, then the intercept column is always added.

So I think there are two questions here:
1. Is the original implementation, which lets users choose if they want an intercept column added or not, out of the scope of the statistic?
2. Why the special casing on the intercept?

For the second question, you're right that the special casing is unnecessary/incorrect. I'm not sure why I append a 1 to the diffs, when it should be a 0 anyhow, and that means that the casing later on is also unnecessary.

If we do want to keep the intercept option in the initialization, some sort of special-casing somewhere will be necessary (at a minimum, adding to the diff ranges). But it could be moved outside of the sensitivity calculation, which I'll do at a minimum. James might be the person to ask about the first question.

Good catch about the `c` assignment; it's my de facto variable for constants and I didn't think about how it would interact with `c()`.

Reply:
I removed special casing and changed `c` assignment. Question 1 still remains unanswered.